Add Danfoss Ally Thermostat

asgothian opened the pull request:
Added Danfoss Ally Thermostat

Koenkk commented on devices.js:
```diff
@@ -8127,7 +8127,7 @@ const devices = [
         supports: 'smoke, tamper and battery',
         fromZigbee: [fz.ias_smoke_alarm_1],
-        toZigbee: [],
+        toZigbee: [tz.warning],
```
Was this added intentionally?

Koenkk commented on devices.js:
```diff
@@ -9922,6 +9922,55 @@ const devices = [
+            tz.danfoss_display_orientation,
+            tz.thermostat_keypad_lockout,
+        ],
+        meta: {configureKey: 3},
```
Please add the required `exposes` property here (used for e.g. Home Assistant discovery). Search this file for examples from other TRVs.

asgothian replied:
Will do, but it will have to wait for the weekend.

sjorge commented on converters/fromZigbee.js:
```diff
@@ -3818,6 +3818,74 @@ const converters = {
+          }
+          if (typeof msg.data['occupiedHeatingSetpoint'] == 'number') {
+              let ohs = precisionRound(msg.data['occupiedHeatingSetpoint'], 2) / 100;
+              // Stelpro will return -325.65 when set to off
```
Seems to be copied from a different device, so maybe update the comment... if it's exactly the same aside from a few values you can also do this... which is cleaner.

https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/converters/fromZigbee.js#L3892
It uses the thermostat_att_report convertor and then adds deviating properties below and fixes up some existing ones that may not be 100% how you expect. It saves on maintenance and code duplication.

asgothian replied:
I'll review the setup on the weekend and verify it works, then update the PR

sjorge commented on devices.js:
```diff
@@ -9922,6 +9922,55 @@ const devices = [
+        description: 'Danfoss Ally Thermostat',
+        supports: 'temperature, heating system control',
+        fromZigbee: [
+           fz.eurotronic_thermostat,
```
You probably don't want this one, it does some ugly shuffling of system_mode and parsing of custom eurotronic bits.

asgothian replied:
I'll review the setup on the weekend and verify it works, then update the PR. I don't want to make any changes without checking the hardware.

asgothian commented:
Fixed all the comments from the review. Note that the exposes are added but not tested as i am not using Home Assistant.

Also added a converter for the Bitron 902010/29 outdoor siren. This was tested earlier today and should work fine. It can only be used to report on battery error or to send an alert to the siren.

Koenkk commented on devices.js:
```diff
@@ -10072,6 +10082,57 @@ const devices = [
+            tz.danfoss_day_of_week, tz.danfoss_trigger_time,
+            tz.danfoss_window_open,
+            tz.danfoss_display_orientation,
+            tz.thermostat_keypad_lockout,
```
Does this thermostat also supports system_mode? (asking because all other thermostats do). Try adding `tz.thermostat_system_mode` and send `{"system_mode": "off"}`, `{"system_mode": "auto"}`,`{"system_mode": "heat"}`

asgothian replied:
I was not able to verify that it does. I could not get it to change to manual or off mode in any reproducible manner

A.

Zeddy- commented:
Hi, the Ally is really slow to respond to setpoint attribute changes, and according to their documentation  you should be able to make it move immediately with this command: 

![image](https://user-images.githubusercontent.com/1147662/99458549-5fab4f00-2935-11eb-9f75-6320be8a6cad.png)

So what i did was added this to zigbee-herdsmans cluster.js hvacthermostat:

hvacThermostat: {
        ID: 513,
        attributes: {
            localTemp: { ID: 0, type: dataType_1.default.int16 },
            outdoorTemp: { ID: 1, type: dataType_1.default.int16 },
            ocupancy: { ID: 2, type: dataType_1.default.bitmap8 },
            absMinHeatSetpointLimit: { ID: 3, type: dataType_1.default.int16 },
            absMaxHeatSetpointLimit: { ID: 4, type: dataType_1.default.int16 },
            absMinCoolSetpointLimit: { ID: 5, type: dataType_1.default.int16 },
            absMaxCoolSetpointLimit: { ID: 6, type: dataType_1.default.int16 },
            pICoolingDemand: { ID: 7, type: dataType_1.default.uint8 },
            pIHeatingDemand: { ID: 8, type: dataType_1.default.uint8 },
            systemTypeConfig: { ID: 9, type: dataType_1.default.bitmap8 },
            localTemperatureCalibration: { ID: 16, type: dataType_1.default.int8 },
            occupiedCoolingSetpoint: { ID: 17, type: dataType_1.default.int16 },
            occupiedHeatingSetpoint: { ID: 18, type: dataType_1.default.int16 },
            unoccupiedCoolingSetpoint: { ID: 19, type: dataType_1.default.int16 },
            unoccupiedHeatingSetpoint: { ID: 20, type: dataType_1.default.int16 },
            minHeatSetpointLimit: { ID: 21, type: dataType_1.default.int16 },
            maxHeatSetpointLimit: { ID: 22, type: dataType_1.default.int16 },
            minCoolSetpointLimit: { ID: 23, type: dataType_1.default.int16 },
            maxCoolSetpointLimit: { ID: 24, type: dataType_1.default.int16 },
            minSetpointDeadBand: { ID: 25, type: dataType_1.default.int8 },
            remoteSensing: { ID: 26, type: dataType_1.default.bitmap8 },
            ctrlSeqeOfOper: { ID: 27, type: dataType_1.default.enum8 },
            systemMode: { ID: 28, type: dataType_1.default.enum8 },
            alarmMask: { ID: 29, type: dataType_1.default.bitmap8 },
            runningMode: { ID: 30, type: dataType_1.default.enum8 },
            startOfWeek: { ID: 32, type: dataType_1.default.enum8 },
            numberOfWeeklyTrans: { ID: 33, type: dataType_1.default.uint8 },
            numberOfDailyTrans: { ID: 34, type: dataType_1.default.uint8 },
            tempSetpointHold: { ID: 35, type: dataType_1.default.enum8 },
            tempSetpointHoldDuration: { ID: 36, type: dataType_1.default.uint16 },
            programingOperMode: { ID: 37, type: dataType_1.default.bitmap8 },
            runningState: { ID: 41, type: dataType_1.default.bitmap16 },
            setpointChangeSource: { ID: 48, type: dataType_1.default.enum8 },
            setpointChangeAmount: { ID: 49, type: dataType_1.default.int16 },
            setpointChangeSourceTimeStamp: { ID: 50, type: dataType_1.default.utc },
            acType: { ID: 64, type: dataType_1.default.enum8 },
            acCapacity: { ID: 65, type: dataType_1.default.uint16 },
            acRefrigerantType: { ID: 66, type: dataType_1.default.enum8 },
            acConpressorType: { ID: 67, type: dataType_1.default.enum8 },
            acErrorCode: { ID: 68, type: dataType_1.default.bitmap32 },
            acLouverPosition: { ID: 69, type: dataType_1.default.enum8 },
            acCollTemp: { ID: 70, type: dataType_1.default.int16 },
            acCapacityFormat: { ID: 71, type: dataType_1.default.enum8 },
            SinopeOccupancy: { ID: 1024, type: dataType_1.default.enum8, manufacturerCode: manufacturerCode_1.default.Sinope },
            SinopeBacklight: { ID: 1026, type: dataType_1.default.enum8, manufacturerCode: manufacturerCode_1.default.Sinope },
            StelproSystemMode: { ID: 0x401c, type: dataType_1.default.enum8 },
            StelproOutdoorTemp: { ID: 0x4001, type: dataType_1.default.int16, manufacturerCode: manufacturerCode_1.default.Stelpro },
	    extWindow: { ID: 16387, type: dataType_1.default.boolean },
        },
        commands: {
            setpointRaiseLower: {
                ID: 0,
                parameters: [
                    { name: 'mode', type: dataType_1.default.uint8 },
                    { name: 'amount', type: dataType_1.default.int8 },
                ],
            },
            setWeeklySchedule: {
                ID: 1,
                parameters: [
                    { name: 'numoftrans', type: dataType_1.default.uint8 },
                    { name: 'dayofweek', type: dataType_1.default.uint8 },
                    { name: 'mode', type: dataType_1.default.uint8 },
                    { name: 'transitions', type: buffaloZclDataType_1.default.LIST_THERMO_TRANSITIONS },
                ],
            },
            getWeeklySchedule: {
                ID: 2,
                parameters: [
                    { name: 'daystoreturn', type: dataType_1.default.uint8 },
                    { name: 'modetoreturn', type: dataType_1.default.uint8 },
                ],
            },
            clearWeeklySchedule: {
                ID: 3,
                parameters: [],
            },
            getRelayStatusLog: {
                ID: 4,
                parameters: [],
            },
            **danfossSetpointCommand: {
		ID: 64,
		parameters: [
                    { name: 'setpointType', type: dataType_1.default.enum8 },
                    { name: 'HeatingSetpoint', type: dataType_1.default.uint16 },
                ],
	    },**
        },
        commandsResponse: {
            getWeeklyScheduleRsp: {
                ID: 0,
                parameters: [
                    { name: 'numoftrans', type: dataType_1.default.uint8 },
                    { name: 'dayofweek', type: dataType_1.default.uint8 },
                    { name: 'mode', type: dataType_1.default.uint8 },
                    { name: 'transitions', type: buffaloZclDataType_1.default.LIST_THERMO_TRANSITIONS },
                ],
            },
            getRelayStatusLogRsp: {
                ID: 1,
                parameters: [
                    { name: 'timeofday', type: dataType_1.default.uint16 },
                    { name: 'relaystatus', type: dataType_1.default.uint16 },
                    { name: 'localtemp', type: dataType_1.default.uint16 },
                    { name: 'humidity', type: dataType_1.default.uint8 },
                    { name: 'setpoint', type: dataType_1.default.uint16 },
                    { name: 'unreadentries', type: dataType_1.default.uint16 },
                ],
            },
        },
    },


**and this to zigbee-herdsman-converters toZigbee.js:**

  hvacThermostat: {
        ID: 513,
        attributes: {
            localTemp: { ID: 0, type: dataType_1.default.int16 },
            outdoorTemp: { ID: 1, type: dataType_1.default.int16 },
            ocupancy: { ID: 2, type: dataType_1.default.bitmap8 },
            absMinHeatSetpointLimit: { ID: 3, type: dataType_1.default.int16 },
            absMaxHeatSetpointLimit: { ID: 4, type: dataType_1.default.int16 },
            absMinCoolSetpointLimit: { ID: 5, type: dataType_1.default.int16 },
            absMaxCoolSetpointLimit: { ID: 6, type: dataType_1.default.int16 },
            pICoolingDemand: { ID: 7, type: dataType_1.default.uint8 },
            pIHeatingDemand: { ID: 8, type: dataType_1.default.uint8 },
            systemTypeConfig: { ID: 9, type: dataType_1.default.bitmap8 },
            localTemperatureCalibration: { ID: 16, type: dataType_1.default.int8 },
            occupiedCoolingSetpoint: { ID: 17, type: dataType_1.default.int16 },
            occupiedHeatingSetpoint: { ID: 18, type: dataType_1.default.int16 },
            unoccupiedCoolingSetpoint: { ID: 19, type: dataType_1.default.int16 },
            unoccupiedHeatingSetpoint: { ID: 20, type: dataType_1.default.int16 },
            minHeatSetpointLimit: { ID: 21, type: dataType_1.default.int16 },
            maxHeatSetpointLimit: { ID: 22, type: dataType_1.default.int16 },
            minCoolSetpointLimit: { ID: 23, type: dataType_1.default.int16 },
            maxCoolSetpointLimit: { ID: 24, type: dataType_1.default.int16 },
            minSetpointDeadBand: { ID: 25, type: dataType_1.default.int8 },
            remoteSensing: { ID: 26, type: dataType_1.default.bitmap8 },
            ctrlSeqeOfOper: { ID: 27, type: dataType_1.default.enum8 },
            systemMode: { ID: 28, type: dataType_1.default.enum8 },
            alarmMask: { ID: 29, type: dataType_1.default.bitmap8 },
            runningMode: { ID: 30, type: dataType_1.default.enum8 },
            startOfWeek: { ID: 32, type: dataType_1.default.enum8 },
            numberOfWeeklyTrans: { ID: 33, type: dataType_1.default.uint8 },
            numberOfDailyTrans: { ID: 34, type: dataType_1.default.uint8 },
            tempSetpointHold: { ID: 35, type: dataType_1.default.enum8 },
            tempSetpointHoldDuration: { ID: 36, type: dataType_1.default.uint16 },
            programingOperMode: { ID: 37, type: dataType_1.default.bitmap8 },
            runningState: { ID: 41, type: dataType_1.default.bitmap16 },
            setpointChangeSource: { ID: 48, type: dataType_1.default.enum8 },
            setpointChangeAmount: { ID: 49, type: dataType_1.default.int16 },
            setpointChangeSourceTimeStamp: { ID: 50, type: dataType_1.default.utc },
            acType: { ID: 64, type: dataType_1.default.enum8 },
            acCapacity: { ID: 65, type: dataType_1.default.uint16 },
            acRefrigerantType: { ID: 66, type: dataType_1.default.enum8 },
            acConpressorType: { ID: 67, type: dataType_1.default.enum8 },
            acErrorCode: { ID: 68, type: dataType_1.default.bitmap32 },
            acLouverPosition: { ID: 69, type: dataType_1.default.enum8 },
            acCollTemp: { ID: 70, type: dataType_1.default.int16 },
            acCapacityFormat: { ID: 71, type: dataType_1.default.enum8 },
            SinopeOccupancy: { ID: 1024, type: dataType_1.default.enum8, manufacturerCode: manufacturerCode_1.default.Sinope },
            SinopeBacklight: { ID: 1026, type: dataType_1.default.enum8, manufacturerCode: manufacturerCode_1.default.Sinope },
            StelproSystemMode: { ID: 0x401c, type: dataType_1.default.enum8 },
            StelproOutdoorTemp: { ID: 0x4001, type: dataType_1.default.int16, manufacturerCode: manufacturerCode_1.default.Stelpro },
	    extWindow: { ID: 16387, type: dataType_1.default.boolean },
        },
        commands: {
            setpointRaiseLower: {
                ID: 0,
                parameters: [
                    { name: 'mode', type: dataType_1.default.uint8 },
                    { name: 'amount', type: dataType_1.default.int8 },
                ],
            },
            setWeeklySchedule: {
                ID: 1,
                parameters: [
                    { name: 'numoftrans', type: dataType_1.default.uint8 },
                    { name: 'dayofweek', type: dataType_1.default.uint8 },
                    { name: 'mode', type: dataType_1.default.uint8 },
                    { name: 'transitions', type: buffaloZclDataType_1.default.LIST_THERMO_TRANSITIONS },
                ],
            },
            getWeeklySchedule: {
                ID: 2,
                parameters: [
                    { name: 'daystoreturn', type: dataType_1.default.uint8 },
                    { name: 'modetoreturn', type: dataType_1.default.uint8 },
                ],
            },
            clearWeeklySchedule: {
                ID: 3,
                parameters: [],
            },
            getRelayStatusLog: {
                ID: 4,
                parameters: [],
            },
            **danfossSetpointCommand: {
		ID: 64,
		parameters: [
                    { name: 'setpointType', type: dataType_1.default.enum8 },
                    { name: 'HeatingSetpoint', type: dataType_1.default.uint16 },
                ],
	    },**
        },
        commandsResponse: {
            getWeeklyScheduleRsp: {
                ID: 0,
                parameters: [
                    { name: 'numoftrans', type: dataType_1.default.uint8 },
                    { name: 'dayofweek', type: dataType_1.default.uint8 },
                    { name: 'mode', type: dataType_1.default.uint8 },
                    { name: 'transitions', type: buffaloZclDataType_1.default.LIST_THERMO_TRANSITIONS },
                ],
            },
            getRelayStatusLogRsp: {
                ID: 1,
                parameters: [
                    { name: 'timeofday', type: dataType_1.default.uint16 },
                    { name: 'relaystatus', type: dataType_1.default.uint16 },
                    { name: 'localtemp', type: dataType_1.default.uint16 },
                    { name: 'humidity', type: dataType_1.default.uint8 },
                    { name: 'setpoint', type: dataType_1.default.uint16 },
                    { name: 'unreadentries', type: dataType_1.default.uint16 },
                ],
            },
        },
    },


**and under devices.js** 

   // Danfoss
     {
	// https://github.com/Koenkk/zigbee-herdsman/blob/master/src/zcl/definition/dataType.ts
        zigbeeModel: ['eTRV0100'],
        model: 'eTRV0100',
        vendor: 'Danfoss',
        description: 'Spirit Zigbee wireless heater thermostat',
        supports: 'temperature, heating system control,battery',
        fromZigbee: [fz.eurotronic_thermostat, fz.battery],
        toZigbee: [
            tz.thermostat_occupied_heating_setpoint,tz.thermostat_extWindow, tz.thermostat_local_temperature_calibration, tz.eurotronic_thermostat_system_mode,
            tz.eurotronic_host_flags, tz.eurotronic_error_status, tz.thermostat_setpoint_raise_lower,**tz.danfoss_setpoint_command**,tz.danfoss_windowopen,
            tz.thermostat_control_sequence_of_operation, tz.thermostat_remote_sensing,tz.danfoss_setpoint_raise_lower,
            tz.eurotronic_current_heating_setpoint, tz.eurotronic_trv_mode, tz.eurotronic_valve_position,
            tz.thermostat_local_temperature,
        ],
        meta: {configureKey: 1},
        ota: ota.zigbeeOTA,
        configure: async (device, coordinatorEndpoint) => {
            const endpoint = device.getEndpoint(1);
            const options = {manufacturerCode: 1246};
            //await bind(endpoint, coordinatorEndpoint, ['hvacThermostat']);
	    await bind(endpoint, coordinatorEndpoint, ['genPowerCfg', 'hvacThermostat']);
            await configureReporting.thermostatTemperature(endpoint, {min: 0, max: repInterval.MINUTES_10, change: 1});
            /*await configureReporting.thermostatPIHeatingDemand(
                endpoint, {min: 0, max: repInterval.MINUTES_5, change: 1},
            );*/
            await configureReporting.thermostatOccupiedHeatingSetpoint(
                endpoint, {min: 0, max: repInterval.MINUTES_5, change: 1},
            );
	    await configureReporting.batteryPercentageRemaining(endpoint);
	    await configureReporting.ctrlSeqeOfOper(endpoint);
	    await configureReporting.thermostatSystemMode(endpoint);
        },
    },

I am able to execute this command, without errors, but it has no effect. Any Ideas? :)

asgothian commented:
Can you link the documentation this was taken from ? I only had access to the cluster definition when building the support. I found that my thermostat reacted reasonably quickly (within 30 seconds) to a set point change. 

Also - I was able to confirm that the thermostat does not accept any mode change commands. The one thing I am still uncertain of is the function of the  “heat available” datapoint.

A.

asgothian commented:
Without this document, i will have to concede that i have done what i can.

How are the chances of this getting merged ?

Koenkk commented on converters/toZigbee.js:
```diff
@@ -1503,6 +1506,132 @@ const converters = {
             }
         },
     },
+    danfoss_mounted_mode: {
```
I think it will be more clean to merge all converters here together, to do this change e.g. `key: ['mounted_mode_control'],` -> `key: ['mounted_mode_control', 'thermostat_orientation'], ` and in the converter do `if (key === 'mounted_mode_control'`,etc.

asgothian replied:
I have had a closer look at that. While it makes the devices.js more readable as one tz. entry replaces a larger number, the single entry in toZigbee becomes a lot more convoluted, as a large number of `if key ===` statements are needed both in `convertGet` and `convertSet`. This is due to the fact that several of the attributes are not marked reportable and will not automatically be reported by the Thermostat unless read explicitly, so the toZigbee includes a `convertGet`

I will set it up like that for all attributes which do not require to be read to verify that they took, but will leave the others as separate entries.

Also - does it make sense to actually read these "non reportable" values during when configuring the reporting to ensure that the system knows their values ? I had not added that, but going through this proposed change made me consider this to be sensible.

asgothian replied:
After reviewing - it is not even feasible to do what i suggested - only one of the affected parameters is marked as reportable. All the others need to be read to confirm their value.

Koenkk commented on devices.js:
```diff
@@ -10080,6 +10090,57 @@ const devices = [
+            tz.thermostat_keypad_lockout,
+        ],
+        exposes: [
+            e.battery(), exposes.climate().withSetpoint('occupied_heating_setpoint', 6, 28, 0.5).withLocalTemperature(),
```
Also add `e.keypad_lockout()`

asgothian replied:
done

asgothian commented on devices.js:
```diff
@@ -10080,6 +10090,57 @@ const devices = [
             await bind(endpoint, coordinatorEndpoint, ['genOnOff', 'genLevelCtrl']);
         },
     },
+    {
```
done
